src/audio/dcblock/dcblock.c

```diff
@@ -0,0 +1,530 @@
+
+	comp_set_drvdata(dev, cd);
+
+	cd->dcblock_func = NULL;
+	cd->config = NULL;
+	cd->config_new = NULL;
+	cd->source_format = 0;
+	cd->sink_format = 0;
+
+	/* Allocate and make a copy of the configu blob. */
+	if (bs) {
+		cd->config = rzalloc(SOF_MEM_ZONE_RUNTIME, 0,
+				     SOF_MEM_CAPS_RAM, bs);
+		if (!cd->config) {
+			rfree(dev);
+			rfree(cd);
+			return NULL;
+		}
+
+		ret = memcpy_s(cd->config, bs, ipc_dcblock->data, bs);
+		assert(!ret);
+	}
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+/**
+ * \brief Frees DC Blocking Filter component.
+ * \param[in,out] dev DC Blocking Filter base component device.
+ */
+static void dcblock_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "dcblock_free()");
+	dcblock_free_config(&cd->config);
+	dcblock_free_config(&cd->config_new);
+	rfree(cd);
+	rfree(dev);
+}
+
+/**
+ * \brief Sets DC Blocking Filter component audio stream parameters.
+ * \param[in,out] dev DC Blocking Filter base component device.
+ * \return Error code.
+ *
+ * All done in prepare() since we need to know source and sink component params.
+ */
+static int dcblock_params(struct comp_dev *dev,
+			  struct sof_ipc_stream_params *params)
+{
+	comp_info(dev, "dcblock_params()");
+
+	return 0;
+}
+
+static int dcblock_cmd_get_data(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata,
+				size_t max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	size_t resp_size;
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "dcblock_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+
+		if (!cd->config) {
+			comp_err(dev, "dcblock_cmd_get_data(), no config");
+			ret = -EINVAL;
+			break;
+		}
+
+		/* Copy config back to user space */
+		resp_size = sizeof(struct sof_dcblock_config);
+		comp_info(dev, "dcblock_cmd_get_data(), resp_size %u",
+			  resp_size);
+
+		if (resp_size > max_size) {
+			comp_err(dev, "response size %i "
+					"exceeds maximum size %i ",
+					resp_size, max_size);
+			ret = -EINVAL;
+			break;
+		}
+
+		ret = memcpy_s(cdata->data->data, cdata->data->size,
+			       cd->config, resp_size);
+		assert(!ret);
+
+		cdata->data->abi = SOF_ABI_VERSION;
+		cdata->data->size = resp_size;
+		break;
+	default:
+		comp_err(dev, "dcblock_cmd_get_data(), invalid command");
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int dcblock_cmd_set_data(struct comp_dev *dev,
+				struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_dcblock_config *request;
+	size_t req_size = sizeof(struct sof_dcblock_config);
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "dcblock_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+
+		/* Retrieve the binary controls from the packet */
+		request = (struct sof_dcblock_config *)cdata->data->data;
+
+		/* Check that there is no work-in-progress previous request */
+		if (!cd->config_new) {
+			comp_err(dev, "dcblock_cmd_set_data(), "
```

Comment:
if we get a new request I would think we would just want to overwrite the old one, unless there is a reason I can't think of.

Reply:
@singalsu Is there a reason the equalizers have this behavior too?

Reply:
The actual coefficients switch is done in EQs in copy() before starting processing to avoid the coefficients to change during calculation. With IIR the in-between coefficients could cause an unstable state, possibly also with this DC blocker IIR. With FIR there's no instability risk but with the other parameters like scaling there could be a short level change glitch instead of clean old -> new coefficients change.

Reply:
That makes sense for IIR, but given we have one value for the config do we need this? The write should have no potential for intermediate state given it is one int per channel.

Reply:
Yes the unknown state is an issue for higher order recursive filters. This should be fine.